### ISSUE TITLE
[Experiment] An alternative choice for sliding window and cyclic shift

### ISSUE THREAD

**Issue by FangGet** [microsoft/Swin-Transformer#65]
First of all, I should point out that it's just a simple experiment for Swin-Transformer.

## Main Idea
As sliding window in Swin-Transformer tries to get correlation attention between different window partitions, we intuitively replace it with pixel shuffle, which is also named as Space2Depth and Depth2Space for Tensorflow. As pixel shuffle does not need window shift, the cyclic shift is also no-more required, thus may give a natural support for non-square window size. What's more, as pixel shuffle splits image in Space2Depth format, one window in W-MSA is connected to all other windows in SW-MSA(though only one pixel is connected for each window), we guess this give a different connection for vit and may gain counterpart or better accuracy.

## Main Change
```
def window_packing(x, window_size):
    B, H, W, C = x.shape
    x= x.contiguous().view(B, window_size, H // window_size, window_size, W // window_size, C)
    windows = x.permute(0, 2, 4, 1, 3, 5).contiguous().view(-1, window_size, window_size, C)
    return windows
```

```
def window_unpacking(windows, window_size, H, W):
    B = int(windows.shape[0] / (H * W / window_size / window_size))
    x = windows.view(B, H // window_size, W // window_size, window_size, window_size, -1)
    x = x.permute(0, 3, 1, 4, 2, 5).contiguous().view(B, H, W, -1)
    return x
```

* Sliding window is removed
* cyclic shift is removed
* no configuration changed

See detail in :
https://github.com/FangGet/Swin-Transformer/blob/main/models/swin_transformer.py

## Result
| name | pretrain | resolution |acc@1 | acc@5 | #params | FLOPs | FPS| 1K model |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |:---: |
| Swin-T | ImageNet-1K | 224x224 | 81.4(+0.2) | 95.6(+0.1) | 28M | 4.5G | 755 | [Google Drive](https://drive.google.com/file/d/1qJaK2G01FZy_m9v6eh0-Sn5Eux8u-qUX/view?usp=sharing)|

**As I have limited GPU resource, if anyone is interested in full experiments for complex architecture and other image tasks, you're welcomed to download the code, run the experiment and report the result.**

## Code
[https://github.com/FangGet/Swin-Transformer](https://github.com/FangGet/Swin-Transformer)

**Comment by huixiancheng** [comment]
nice work.Mean that we don't need to calculate the atten mask then full with ”inf“?  
I also try to use non-square window to finish my task, look like your implement much better and easy.

**Comment by huixiancheng** [comment]
Have you check this part? https://github.com/microsoft/Swin-Transformer/blob/bc26f4764735a91b0330071cc9ca9ffcf60340fa/models/swin_transformer.py#L192-L196.
In orginal paper, they keep two W-MSA in last stage since window_size == feature_resolution. 
Maybe should keep **do_ps == False** when min(self.input_resolution) <= self.window_size for fair comparison.
Though in sem.seg, they don't do as so. https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/issues/18#issuecomment-837990862

**Comment by FangGet** [comment]
For pixel shuffle, SW-MSA seems to degrade to W-MSA when window_size == feature_resolution? though the additional operation is also a problem. 
Just a thought, no strict experiment. I will check it later.
